### PR TITLE
Update Api end points and add brs for allowlist

### PR DESCRIPTION
**Pull request by kcheekuri**
[Preview this branch](https://ms.portal.azure.com/?feature.phoenixendpoint=https://juno-test.documents-dev.windows-int.net&dataExplorerSource=https://cosmos-explorer-preview.azurewebsites.net/commit/288c85d13ce232ee1aac50cded93f93e27ab7008/explorer.html?feature.pr=https://github.com/Azure/cosmos-explorer/pull/1161#home)

This PR:
- Updated Api end points for allocate, reset and container health check.
- For phoenix feature, allows white listed accounts through Api + phoenix  feature flag. 
- Disable phoenix for vnet/firewall enabled accounts.
- Terminal is hidden behind feature flag and MPAC.

### CODE REVIEW

File: src/Phoenix/PhoenixClient.ts [hunk 1]
```diff
@@ -130,7 +181,7 @@ export class PhoenixClient {
   }
 
   public getPhoenixContainerPoolingEndPoint(): string {
```
Comment (srinathnarayanan) [inline src/Phoenix/PhoenixClient.ts:183]: Remove AccountData type, you can get the path using 
`${this.getPhoenixContainerPoolingEndPoint()}{userContext.databaseAccount.id}/containerconnections`
Make

Comment (srinathnarayanan) [inline src/Phoenix/PhoenixClient.ts:183]: getPhoenixControlPlaneEndpoint

File: src/Contracts/DataModels.ts [hunk 2]
```diff
@@ -438,14 +438,16 @@ export interface ContainerInfo {
 }
 
 export interface IProvisionData {
+  cosmosEndpoint: string;
+}
+
+export interface IAccountData {
```
Comment (srinathnarayanan) [inline src/Contracts/DataModels.ts:444]: Remove this

File: src/Explorer/Explorer.tsx [hunk 3]
```diff
@@ -412,9 +403,11 @@ export default class Explorer {
     connectionStatus: DataModels.ContainerConnectionInfo
   ) {
     if (connectionInfo.status === HttpStatusCodes.OK && connectionInfo.data && connectionInfo.data.notebookServerUrl) {
-      const containerData = {
-        forwardingId: connectionInfo.data.forwardingId,
+      const containerData: IContainerData = {
```
Comment (srinathnarayanan) [inline src/Explorer/Explorer.tsx:406]: only forwardingid is needed

File: src/Explorer/Explorer.tsx [hunk 4]
```diff
@@ -1292,17 +1284,21 @@ export default class Explorer {
       : this.refreshAllDatabases();
     await useNotebook.getState().refreshNotebooksEnabledStateForAccount();
 
+    //Disable phoenix in case of Vnet or Firewall was enabled.
+    if (useNotebook.getState().isPhoenix && !isPublicInternetAccessAllowed()) {
```
Comment (srinathnarayanan) [inline src/Explorer/Explorer.tsx:1288]: remove useNotebook.getState().isPhoenix check

File: src/Explorer/Explorer.tsx [hunk 5]
```diff
@@ -1292,17 +1284,21 @@ export default class Explorer {
...
-    useNotebook.getState().setIsShellEnabled(userContext.features.phoenix && isPublicInternetAccessAllowed());
+    useNotebook.getState().setIsShellEnabled(useNotebook.getState().isPhoenix && isPublicInternetAccessAllowed());
 
     TelemetryProcessor.trace(Action.NotebookEnabled, ActionModifiers.Mark, {
       isNotebookEnabled,
       dataExplorerArea: Constants.Areas.Notebook,
     });
 
-    if (!userContext.features.notebooksTemporarilyDown) {
+    if (useNotebook.getState().isPhoenix) {
```
Comment (srinathnarayanan) [inline src/Explorer/Explorer.tsx:1301]: simplyfy, use only useNoteboo.gettate().isPhoenix check

File: src/Explorer/Notebook/NotebookContainerClient.ts [hunk 6]
```diff
@@ -131,21 +131,29 @@ export class NotebookContainerClient {
...
         throw new AbortError(response.statusText);
       }
-      throw new Error();
+      throw new Error(response.statusText);
     } else {
       return undefined;
     }
   }
 
   private checkStatus(): boolean {
```
Comment (srinathnarayanan) [inline src/Explorer/Notebook/NotebookContainerClient.ts:140]: change to shouldExecuteMemoryCall

File: src/Explorer/Notebook/NotebookContainerClient.ts [hunk 7]
```diff
@@ -131,21 +131,29 @@ export class NotebookContainerClient {
...
-    if (NotebookUtil.isPhoenixEnabled()) {
-      if (useNotebook.getState().containerStatus?.status === Constants.ContainerStatusType.Disconnected) {
+    if (useNotebook.getState().isPhoenix) {
```
Comment (srinathnarayanan) [inline src/Explorer/Notebook/NotebookContainerClient.ts:141]: check when you should return true and do only that check, the resetting of other states should be done based on the health check, not the memory call

File: src/Explorer/Tree/ResourceTree.tsx [hunk 8]
```diff
@@ -121,7 +121,7 @@ export const ResourceTree: React.FC<ResourceTreeProps> = ({ container }: Resourc
       children: [],
     };
 
-    if (userContext.features.notebooksTemporarilyDown) {
+    if (useNotebook.getState().isPhoenix === false) {
```
Comment (srinathnarayanan) [inline src/Explorer/Tree/ResourceTree.tsx:124]: !useNotebook.getState().isPhoenix

File: src/Explorer/Notebook/NotebookContainerClient.ts (hunk 7 above)
Comment (srinathnarayanan) [reply src/Explorer/Notebook/NotebookContainerClient.ts:141]: Remove setting/resetting state in scheduleHeartbeat as well

File: src/Phoenix/PhoenixClient.ts [hunk 9]
```diff
@@ -55,7 +69,7 @@ export class PhoenixClient {
     }
   }
 
-  public async initiateContainerHeartBeat(containerData: { forwardingId: string; dbAccountName: string }) {
+  public async initiateContainerHeartBeat(containerData: IContainerData) {
```
Comment (srinathnarayanan) [inline src/Phoenix/PhoenixClient.ts:72]: remove try/catch in getContainerHealth

File: src/Explorer/Notebook/NotebookContainerClient.ts [hunk 10]
```diff
@@ -202,25 +175,6 @@ export class NotebookContainerClient {
     };
   }
 
-  private async recreateNotebookWorkspaceAsync(): Promise<void> {
```
Comment (srinathnarayanan) [inline src/Explorer/Notebook/NotebookContainerClient.ts:205]: why was this removed?

File: src/Explorer/SplashScreen/SplashScreen.tsx [hunk 11]
```diff
@@ -84,7 +84,7 @@ export class SplashScreen extends React.Component<SplashScreenProps> {
     const mainItems = this.createMainItems();
     const commonTaskItems = this.createCommonTaskItems();
     let recentItems = this.createRecentItems();
-    if (userContext.features.notebooksTemporarilyDown) {
+    if (useNotebook.getState().isPhoenix) {
```
Comment (srinathnarayanan) [inline src/Explorer/SplashScreen/SplashScreen.tsx:87]: should this be "!useNotebook.getState().isPhoenix"

File: src/Phoenix/PhoenixClient.ts [hunk 12]
```diff
@@ -72,7 +76,7 @@ export class PhoenixClient {
     try {
       const runContainerStatusAsync = async () => {
         const response = await window.fetch(
-          `${this.getPhoenixContainerPoolingEndPoint()}/${containerData.dbAccountName}/${containerData.forwardingId}`,
+          `${this.getPhoenixControlPlaneEndpoint()}${userContext.databaseAccount.id}/${containerData.forwardingId}`,
```
Comment (srinathnarayanan) [inline src/Phoenix/PhoenixClient.ts:79]: move using "${userContext.databaseAccount.id}" to getPhoenixControlPlaneEndpoint itself.

File: src/Explorer/SplashScreen/SplashScreen.tsx (hunk 11 above)
Comment (kcheekuri) [reply src/Explorer/SplashScreen/SplashScreen.tsx:87]: No, according to previous flag when notebooks are down then we need to remove 'Notebooks' from recent activity. Now with this flag we need only when phoenix enabled. Removing this check so that default removes notebooks from recent activity

File: src/Explorer/Notebook/NotebookContainerClient.ts (hunk 10 above)
Comment (kcheekuri) [reply src/Explorer/Notebook/NotebookContainerClient.ts:205]: Old compute v2 code

**Review by srinathnarayanan — DISMISSED** [review]

**Review by srinathnarayanan — APPROVED** [review]

File: configs/mpac.json [hunk 13]
```diff
@@ -1,3 +1,4 @@
 {
-  "JUNO_ENDPOINT": "https://tools-staging.cosmos.azure.com"
+  "JUNO_ENDPOINT": "https://tools-staging.cosmos.azure.com",
+  "IS_MPAC" : true
```
Comment (swviswan) [inline configs/mpac.json:3]: >IS_MPAC [](http://example.com/codeflow?start=3&length=7)

IsTerminalEnabled

File: src/Explorer/Notebook/useNotebook.ts [hunk 14]
```diff
@@ -292,4 +297,12 @@ export const useNotebook: UseStore<NotebookState> = create((set, get) => ({
   },
   setIsRefreshed: (isRefreshed: boolean) => set({ isRefreshed }),
   setContainerStatus: (containerStatus: ContainerInfo) => set({ containerStatus }),
+  getPhoenixStatus: async () => {
+    if (get().isPhoenix === undefined) {
+      const phoenixClient = new PhoenixClient();
```
Comment (swviswan) [inline src/Explorer/Notebook/useNotebook.ts:302]: >PhoenixClient [](http://example.com/codeflow?start=32&length=13)

dont need to create it if usercontext.features.phoenix is false.

**Comment by swviswan** [comment]
>         useNotebook.getState().setConnectionInfo(connectionStatus);

use local variable within the fn instead of callig useNotebooks.getstate multiple times within the same fn

---
Refers to: src/Explorer/Explorer.tsx:524 in a9b67bc. [](commit_id = a9b67bcc1b0f7193d79449a271b3bda389e04576, deletion_comment = False)

File: src/Explorer/Explorer.tsx [hunk 15]
```diff
@@ -1303,23 +1286,22 @@ export default class Explorer {
...
+    //Disable phoenix in case of Vnet or Firewall was enabled.
+    if (!isPublicInternetAccessAllowed()) {
+      useNotebook.getState().setIsPhoenix(false);
+    }
     // TODO: remove reference to isNotebookEnabled and isNotebooksEnabledForAccount
-    const isNotebookEnabled = userContext.features.notebooksDownBanner || userContext.features.phoenix;
+    const isNotebookEnabled = userContext.features.notebooksDownBanner || useNotebook.getState().isPhoenix;
     useNotebook.getState().setIsNotebookEnabled(isNotebookEnabled);
-    useNotebook.getState().setIsShellEnabled(userContext.features.phoenix && isPublicInternetAccessAllowed());
+    useNotebook.getState().setIsShellEnabled(useNotebook.getState().isPhoenix && isPublicInternetAccessAllowed());
```
Comment (swviswan) [inline src/Explorer/Explorer.tsx:1296]: >useNotebook.getState().isPhoenix [](http://example.com/codeflow?start=45&length=32)

setIshellenabled = isPublicInternetAccessAllowed()

File: src/Explorer/ContextMenuButtonFactory.tsx [hunk 16]
```diff
@@ -83,7 +83,7 @@ export const createCollectionContextMenuButton = (
 
     items.push({
       iconSrc: HostedTerminalIcon,
-      isDisabled: useNotebook.getState().isShellEnabled && userContext.features.notebooksTemporarilyDown,
+      isDisabled: useNotebook.getState().isShellEnabled && !useNotebook.getState().isPhoenix,
```
Comment (swviswan) [inline src/Explorer/ContextMenuButtonFactory.tsx:86]: >useNotebook.getState().isShellEnabled && !useNotebook.getState().isPhoenix, [](http://example.com/codeflow?start=18&length=75)

!isShellEnabled

File: src/Explorer/Explorer.tsx [hunk 17]
```diff
@@ -1303,23 +1286,22 @@ export default class Explorer {
       : this.refreshAllDatabases();
     await useNotebook.getState().refreshNotebooksEnabledStateForAccount();
 
+    //Disable phoenix in case of Vnet or Firewall was enabled.
+    if (!isPublicInternetAccessAllowed()) {
+      useNotebook.getState().setIsPhoenix(false);
```
Comment (swviswan) [inline src/Explorer/Explorer.tsx:1291]: move it to wherever isPhoenix is defined in useNotebooks.ts file..

Alyways - one place for all logic is better.

File: src/Explorer/Menus/CommandBar/CommandBarComponentButtonFactory.tsx [hunk 18]
```diff
@@ -179,12 +180,12 @@ export function createContextCommandBarButtons(
       ariaLabel: label,
       hasPopup: true,
       tooltipText:
-        useNotebook.getState().isShellEnabled && userContext.features.notebooksTemporarilyDown
+        useNotebook.getState().isShellEnabled && !useNotebook.getState().isPhoenix
           ? Constants.Notebook.mongoShellTemporarilyDownMsg
```
Comment (swviswan) [inline src/Explorer/Menus/CommandBar/CommandBarComponentButtonFactory.tsx:184]: not blocking checkin: please check with Srinath/Tanuj if this  is needed.

File: src/Explorer/Menus/CommandBar/CommandBarComponentButtonFactory.tsx [hunk 19]
```diff
@@ -179,12 +180,12 @@ export function createContextCommandBarButtons(
...
           : undefined,
       disabled:
         (selectedNodeState.isDatabaseNodeOrNoneSelected() && userContext.apiType === "Mongo") ||
-        (useNotebook.getState().isShellEnabled && userContext.features.notebooksTemporarilyDown),
+        (useNotebook.getState().isShellEnabled && !useNotebook.getState().isPhoenix),
```
Comment (swviswan) [inline src/Explorer/Menus/CommandBar/CommandBarComponentButtonFactory.tsx:188]: >(useNotebook.getState().isShellEnabled && !useNotebook.getState().isPhoenix [](http://example.com/codeflow?start=8&length=75)

may be we need to remove this

**Comment by swviswan** [comment]
>       set({ isNotebooksEnabledForAccount: false });

not blocking checkin: can u evaluate if this needs ot be cleanedup

---
Refers to: src/Explorer/Notebook/useNotebook.ts:118 in a9b67bc. [](commit_id = a9b67bcc1b0f7193d79449a271b3bda389e04576, deletion_comment = False)

File: src/Phoenix/PhoenixClient.ts [hunk 20]
```diff
@@ -129,8 +142,10 @@ export class PhoenixClient {
     return phoenixEndpoint;
   }
 
-  public getPhoenixContainerPoolingEndPoint(): string {
-    return `${PhoenixClient.getPhoenixEndpoint()}/api/controlplane/toolscontainer`;
+  public getPhoenixControlPlaneEndpoint(): string {
```
Comment (swviswan) [inline src/Phoenix/PhoenixClient.ts:145]: rename to ControlPlanePathPrefix or something like that

File: src/Phoenix/PhoenixClient.ts [hunk 21]
```diff
@@ -86,13 +84,23 @@ export class PhoenixClient {
             status: ContainerStatusType.Active,
           };
         } else if (response.status === HttpStatusCodes.NotFound) {
```
Comment (swviswan) [inline src/Phoenix/PhoenixClient.ts:86]: add log for this case

**Review by swviswan — DISMISSED** [review]
:shipit:

File: src/Explorer/Explorer.tsx (hunk 17 above)
Comment (kcheekuri) [reply src/Explorer/Explorer.tsx:1291]: Removed this in current PR

File: src/Phoenix/PhoenixClient.ts [hunk 22]
```diff
@@ -86,13 +96,26 @@ export class PhoenixClient {
...
+          TelemetryProcessor.traceMark(Action.PhoenixReconnect, {
+            dataExplorerArea: Areas.Notebook,
```
Comment (swviswan) [inline src/Phoenix/PhoenixClient.ts:100]: >dataExplorerArea [](http://example.com/codeflow?start=12&length=16)

would this have db account and puid so we can get distinct count of user/accounts having running into this issue

Comment (swviswan) [reply src/Phoenix/PhoenixClient.ts:100]: Also - as informed, we  need log for when reconnect happens.

File: src/Phoenix/PhoenixClient.ts [hunk 23]
```diff
@@ -86,13 +96,26 @@ export class PhoenixClient {
...
+          };
+          useNotebook.getState().resetContainerConnection(connectionStatus);
+          useNotebook.getState().setIsRefreshed(!useNotebook.getState().isRefreshed);
           throw new AbortError(response.statusText);
         }
         throw new Error(response.statusText);
       };
       return await promiseRetry(runContainerStatusAsync, this.retryOptions);
     } catch (error) {
       Logger.logError(getErrorMessage(error), "PhoenixClient/getContainerStatus");
```
Comment (swviswan) [inline src/Phoenix/PhoenixClient.ts:113]: >logError [](http://example.com/codeflow?start=13&length=8)

why only logerror and not telemetryprocessor used here?

File: src/Explorer/ContextMenuButtonFactory.tsx [hunk 24]
```diff
@@ -83,7 +83,6 @@ export const createCollectionContextMenuButton = (
 
     items.push({
       iconSrc: HostedTerminalIcon,
-      isDisabled: useNotebook.getState().isShellEnabled && userContext.features.notebooksTemporarilyDown,
       onClick: () => {
         const selectedCollection: ViewModels.Collection = useSelectedNode.getState().findSelectedCollection();
         if (useNotebook.getState().isShellEnabled) {
```
Comment (swviswan) [inline src/Explorer/ContextMenuButtonFactory.tsx:88]: >isShellEnabled [](http://example.com/codeflow?start=35&length=14)

isShellEnabled or isPhoenix?

File: src/Phoenix/PhoenixClient.ts (hunk 22 above)
Comment (kcheekuri) [reply src/Phoenix/PhoenixClient.ts:100]: > > dataExplorerArea
> 
> would this have db account and puid so we can get distinct count of user/accounts having running into this issue

Yes the record will have the details

File: src/Phoenix/PhoenixClient.ts [hunk 25]
```diff
@@ -71,28 +81,53 @@ export class PhoenixClient {
   private async getContainerStatusAsync(containerData: IContainerData): Promise<ContainerInfo> {
     try {
       const runContainerStatusAsync = async () => {
+        TelemetryProcessor.traceStart(Action.PhoenixHeartBeat, {
+          dataExplorerArea: Areas.Notebook,
```
Comment (swviswan) [inline src/Phoenix/PhoenixClient.ts:85]: why do we need ths telemetry..
for 1 user - this will be every few sec/minutes.. and will beso noisy.. do we gain benefit by having this?

File: src/Phoenix/PhoenixClient.ts [hunk 26]
```diff
@@ -71,28 +81,53 @@ export class PhoenixClient {
...
+          });
           const containerStatus = await response.json();
           return {
             durationLeftInMinutes: containerStatus?.durationLeftInMinutes,
             notebookServerInfo: containerStatus?.notebookServerInfo,
             status: ContainerStatusType.Active,
           };
         } else if (response.status === HttpStatusCodes.NotFound) {
+          const error = "Container got disconnected from the workspace";
+          Logger.logError(error, "PhoenixClient/getContainerStatusAsync");
```
Comment (swviswan) [inline src/Phoenix/PhoenixClient.ts:106]: >Logger.logError [](http://example.com/codeflow?start=10&length=15)

do we need this?
guess elemetry processor will send the error to appinsights as well whhich is what logerror does?

File: src/Phoenix/PhoenixClient.ts (hunk 25 above)
Comment (kcheekuri) [reply src/Phoenix/PhoenixClient.ts:85]: Added this if we want to know the rate of heart beat success

**Review by swviswan — APPROVED** [review]
:shipit:

File: src/Phoenix/PhoenixClient.ts (hunk 26 above)
Comment (kcheekuri) [reply src/Phoenix/PhoenixClient.ts:106]: This is the logger that gets displayed in portal console

Comment (swviswan) [reply src/Phoenix/PhoenixClient.ts:100]: i dont think we want that.. console logs should be only for development

**Review by swviswan — APPROVED** [review]
:shipit:

File: src/Explorer/Notebook/useNotebook.ts [hunk 27]
```diff
@@ -196,7 +201,7 @@ export const useNotebook: UseStore<NotebookState> = create((set, get) => ({
     isGithubTree ? set({ gitHubNotebooksContentRoot: root }) : set({ myNotebooksContentRoot: root });
   },
   initializeNotebooksTree: async (notebookManager: NotebookManager): Promise<void> => {
-    const notebookFolderName = NotebookUtil.isPhoenixEnabled() === true ? "Temporary Notebooks" : "My Notebooks";
+    const notebookFolderName = get().isPhoenix === true ? "Temporary Notebooks" : "My Notebooks";
```
Comment (languy) [inline src/Explorer/Notebook/useNotebook.ts:204]:
```suggestion
    const notebookFolderName = get().isPhoenix ? "Temporary Notebooks" : "My Notebooks";
```

File: src/Explorer/Notebook/useNotebook.ts [hunk 28]
```diff
@@ -292,4 +297,15 @@ export const useNotebook: UseStore<NotebookState> = create((set, get) => ({
   },
   setIsRefreshed: (isRefreshed: boolean) => set({ isRefreshed }),
   setContainerStatus: (containerStatus: ContainerInfo) => set({ containerStatus }),
+  getPhoenixStatus: async () => {
+    if (get().isPhoenix === undefined) {
+      let isPhoenix = false;
+      if (userContext.features.phoenix === true) {
```
Comment (languy) [inline src/Explorer/Notebook/useNotebook.ts:303]:
```suggestion
      if (userContext.features.phoenix) {
```

File: src/Phoenix/PhoenixClient.ts [hunk 29]
```diff
@@ -86,13 +96,32 @@ export class PhoenixClient {
             status: ContainerStatusType.Active,
           };
         } else if (response.status === HttpStatusCodes.NotFound) {
+          const error = "Disconnected from compute workspace";
+          Logger.logError(error, "PhoenixClient/ContainerStatus");
```
Comment (swviswan) [inline src/Phoenix/PhoenixClient.ts:100]: >PhoenixClient/ContainerStatus [](http://example.com/codeflow?start=34&length=29)

dont need this - can u pas empty string here?

for customer - PhoenixClient is not something meaningful.

**Review by swviswan — DISMISSED** [review]
:shipit:

File: src/Phoenix/PhoenixClient.ts [hunk 30]
```diff
@@ -101,19 +130,24 @@ export class PhoenixClient {
...
-  private async getContainerHealth(delayMs: number, containerData: { forwardingId: string; dbAccountName: string }) {
+  private async getContainerHealth(delayMs: number, containerData: IContainerData) {
+    const containerInfo = await this.getContainerStatusAsync(containerData);
+    useNotebook.getState().setContainerStatus(containerInfo);
+    if (useNotebook.getState().containerStatus?.status === ContainerStatusType.Active) {
+      this.scheduleContainerHeartbeat(delayMs, containerData);
+    }
+  }
+
+  public async IsDbAcountWhitelisted() {
```
Comment (languy) [inline src/Phoenix/PhoenixClient.ts:141]: Unlike C# which is PascalCase, typescript is camelCase. Add return type.
```suggestion
  public async isDbAcountWhitelisted(): Promise<boolean> {
```

**Review by languy — APPROVED** [review]

**Review by languy — APPROVED** [review]
